fs: enable fstab devicetree integration for fatfs

kica-z opened the pull request:
In order to use fstab with fatfs in addition to littlefs some changes where introduces in the fatfs integration of zephyr.
Example Usage:
```c
/ {
    fstab {
        compatible = "zephyr,fstab";
        ffs1: ffs1 {
            compatible = "zephyr,fstab,fatfs";
            automount;
            disk-access;
            mount-point = "/SD";
        };
    };
};
```

de-nordic commented on subsys/fs/fat_fs.c:
```diff
@@ -557,10 +557,52 @@ static const struct fs_file_system_t fatfs_fs = {
+	}
 
-	return fs_register(FS_FATFS, &fatfs_fs);
+	return rc;
```
This will be a little bit of problem, because failing to mount any fstab, for example due to damaged fat, will imply that registration of fs failed.

There is also needed a testcase for automount of fat, and this PR does not provide any.

kica-z replied:
I will remove the return value overwrite so that the return only indicates if the registration of the fs succeeded or not.

I will also provided a testcase.

decsny commented on subsys/fs/fat_fs.c:
```diff
@@ -557,10 +557,53 @@ static const struct fs_file_system_t fatfs_fs = {
+
 static int fatfs_init(void)
 {
+	int rc = fs_register(FS_FATFS, &fatfs_fs);
+
+	if (rc == 0) {
+		static struct fs_mount_t *partitions[] = {
+			DT_INST_FOREACH_STATUS_OKAY(REFERENCE_MOUNT)};
+		int ret = 0;
+
+		for (size_t i = 0; i < ARRAY_SIZE(partitions); i++) {
+			struct fs_mount_t *mpi = partitions[i];
+
+			if ((mpi->flags & FS_MOUNT_FLAG_AUTOMOUNT) != 0) {
+				ret = fs_mount(mpi);
+				if (ret < 0) {
+					LOG_ERR("Error mounting filesystem: at %s: %d",
+						mpi->mnt_point, ret);
+				} else {
+					LOG_DBG("FATFS Filesystem \"%s\" initialized",
+						mpi->mnt_point);
+				}
+			}
+		}
+	}
```
there is way too much nesting here making this code hard to understand, split into helper functions please (I don't maintain this FS code though so if the maintainers disagree with this suggestion then ok)

kica-z replied:
I adjusted the code to reduce nesting.

kica-z commented on dts/bindings/fs/zephyr,fstab,littlefs.yaml:
```diff
@@ -11,6 +11,12 @@ include: "zephyr,fstab-common.yaml"
 properties:
   # num-files and num-dirs are not filesystem-specific.
 
+  partition:
+    type: phandle
+    required: true
+    description: |
+      A reference to the file system's partition.
```
Fair enough i adjusted the commit message.

The property was moved as it is currently just used by littlefs and it would be strange to have to define it for fatfs even though it is not used. I am open to alternatives should there be better solutions for this.

kica-z commented on dts/bindings/fs/zephyr,fstab,fatfs.yaml:
```diff
@@ -0,0 +1,9 @@
+description: |
+    Description of pre-defined fatfs file systems.
+
+compatible: "zephyr,fstab,fatfs"
```
Any news on this?

decsny replied:
don't worry about this for now in the PR since it is consistent with what is in the tree

rruuaanng replied:
Sorry, @decsny. I've been a bit busy recently and didn't notice this message. I will answer the question later, and left a footprint here.

rruuaanng replied:
From the EPAPR perspective, they are just "specifications". The compatible property can be any value, such as "abc" or "foo." Even though we encourage the use of `<"manufacturer,model">`, if new descriptions can make it more precise, that would be better. So, I have no objections to this.

de-nordic commented on subsys/fs/fat_fs.c:
```diff
@@ -557,10 +557,56 @@ static const struct fs_file_system_t fatfs_fs = {
+	BUILD_ASSERT(!DT_INST_PROP(inst, no_format),                                               \
+		     "NO_FORMAT not supported for individual instanzes FS_FATFS_MKFS");            \
+	FATFS fs_data_##inst;                                                                      \
+	struct fs_mount_t FS_FSTAB_ENTRY(DT_DRV_INST(inst)) = {                                    \
```
Shouldn't this be static?

kica-z replied:
You are right it should be. I just took this part from littlefs without checking linkage (it is the same there).
I have adjusted this.

kica-z commented on subsys/fs/fat_fs.c:
```diff
@@ -557,10 +557,56 @@ static const struct fs_file_system_t fatfs_fs = {
+
+			automount_if_enabled(mpi);
+		}
+	}
```
I added a kconfig for this.

kica-z commented on subsys/fs/fat_fs.c:
```diff
@@ -557,10 +557,56 @@ static const struct fs_file_system_t fatfs_fs = {
+	int rc = fs_register(FS_FATFS, &fatfs_fs);
+
+	if (rc == 0) {
+		struct fs_mount_t *partitions[] = {DT_INST_FOREACH_STATUS_OKAY(REFERENCE_MOUNT)};
```
I added a devicetree check so that the code is not executed if there is no instance of the driver.

kica-z commented on subsys/fs/fat_fs.c:
```diff
@@ -557,6 +557,45 @@ static const struct fs_file_system_t fatfs_fs = {
 #endif
 };
 
+#define DT_DRV_COMPAT zephyr_fstab_fatfs
```
@de-nordic I have created a macro for this. I think it would make sense to upstream this. What do you think?
```c
#define ZZZ_PROP_CHECK(inst, prop) COND_CODE_1(DT_PROP(DT_DRV_INST(inst), prop), (1), ())
#define DT_DRV_ANY_OKAY_PROP(prop)                                                                 \
	COND_CODE_1(IS_EMPTY(DT_INST_FOREACH_STATUS_OKAY_VARGS(AUTOMOUNT_CHECK2, prop)), (0), (1))

#if DT_DRV_ANY_OKAY_PROP(automount);
#error A driver instance has the property automount set
#endif
```

kica-z replied:
@de-nordic what do you think of this?

de-nordic replied:
This is what I mean: https://github.com/zephyrproject-rtos/zephyr/pull/84147/files#r1969586710
This will make the Kconfig option to not appear when there is no enabled dts entry for the FAT fstab.

de-nordic replied:
Then you do not have to do the 
```
#ifdef CONFIG_FS_FATFS_FSTAB_AUTOMOUNT
#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
```
because the `CONFIG_FS_FATFS_FSTAB_AUTOMOUNT=y` would both confirm that DTS entry exists and user wants the feature on.

kica-z replied:
@de-nordic but isn't the macro better because the kconfig is no longer necessary? The feature is just there when needed and not otherwise?

kica-z replied:
@de-nordic  Ok. I think this is a preference question but i see your point and your suggestion also is inline with the rest of the zephyr-ecosystem. I adjusted accordingly

de-nordic commented on subsys/fs/Kconfig.fatfs:
```diff
@@ -299,6 +299,12 @@ config FS_FATFS_CUSTOM_MOUNT_POINTS
 	   be used for mounting fatfs filesystems anymore.
 	depends on FS_FATFS_CUSTOM_MOUNT_POINT_COUNT > 0
 
+config FS_FATFS_FSTAB_AUTOMOUNT
+	bool "Support for fstab auto-mounting"
+	help
+	  This option enables fatfs support for the devicetree fstab auto-mounting
+	  functionality.
+
```
```suggestion
DT_COMPAT_ZEPHYR_FSTAB_FATFS := zephyr,fstab,fatfs

config FS_FATFS_FSTAB_AUTOMOUNT
	bool "Support for fstab auto-mounting"
	depends on $(dt_compat_enabled, $(DT_COMPAT_ZEPHYR_FSTAB_FATFS))
	help
	  This option enables fatfs support for the devicetree fstab auto-mounting
	  functionality.

```

clamattia commented:
Looking forward to this.

kica-z commented on subsys/fs/fat_fs.c:
```diff
@@ -557,6 +557,43 @@ static const struct fs_file_system_t fatfs_fs = {
+		.flags = FSTAB_ENTRY_DT_MOUNT_FLAGS(DT_DRV_INST(inst)),                            \
+	};
+
+DT_INST_FOREACH_STATUS_OKAY(DEFINE_FS);
```
@de-nordic i adjusted this. But shouldn't the fstab entry still be created even if automount is disabled symantically speaking? Wouldn't the developer expect a FS_FSTAB_ENTRY to exist? (i don't know how this is handled normally that's why i am asking)

de-nordic replied:
You only need to list these here in case when you do automount.
If there is not automount, there is no need to store these object here as there is no usage fro them, and users actually have no way to pick them up.
Users should be able to define FS from fstab using this:
https://github.com/zephyrproject-rtos/zephyr/blob/b74b092f22ea29950f291104a84e3433f4efd7d6/include/zephyr/fs/fs.h#L215
and that should give them the object that then they can pass to fs_mount.

kica-z replied:
@de-nordic ok, so we can leave it like it is now within the ifdef, correct? Could you then approve it again? The approval got lost since i had to push again :)

de-nordic replied:
Give me a moment, I might be wrong. I have pulled down your PR and am running to check some things.

de-nordic replied:
OK, so I was wrong, here is PR to your PR https://github.com/kica-z/zephyr/pull/1, that makes fstab entries available to users, it removes the static from fstab object definitions and moves the ifdef back to where it is supposed to be; and there is also modification of your sample so that the direct usage of fstab is show.

de-nordic replied:
Two notes for the future (so I do not request changes in the pr):
 1) I have tested this by hand and via sample, but we need to write a test for fstab.
 2) the sample you have provided could be made universal to show littlefs mounting too.

kica-z replied:
> OK, so I was wrong, here is PR to your PR [kica-z#1](https://github.com/kica-z/zephyr/pull/1), that makes fstab entries available to users, it removes the static from fstab object definitions and moves the ifdef back to where it is supposed to be; and there is also modification of your sample so that the direct usage of fstab is show.

@de-nordic LGTM. I also did not know that that is how to access FSTAB entries the "traditional" way 😅. Thanks for creating the PR. I integrated it into my commits and credited you i hope that is ok.

kica-z replied:
All good. Thanks for the help 👍🙂

kica-z committed:
fs: enable fstab devicetree integration for fatfs

In order to use fstab with fatfs in addition to
littlefs a new driver fstab,fatfs was introduced
containing the required logic to detect and if
needed automount fatfs devices specified in fstab.
Additionally as the partition phandle is not needed
by fatfs and currently is specific to littlefs it
was moved from the common dts binding into the
littlefs binding.

Signed-off-by: Carlo Kirchmeier <carlo.kirchmeier@zuehlke.com>
Co-authored-by: Carlo Kirchmeier <carlo.kirchmeier@zuehlke.com>
Co-authored-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

kica-z committed:
test: fs: enable fstab devicetree integration for fatfs

Added tests in order to verify the functionality of the
newly added fatfs fstab integration.

Signed-off-by: Carlo Kirchmeier <carlo.kirchmeier@zuehlke.com>
Co-authored-by: Carlo Kirchmeier <carlo.kirchmeier@zuehlke.com>
Co-authored-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic reviewed (approved):
Looks OK, thanks!

kica-z commented:
@de-nordic the pipeline seems to be fixed could you maybe rerun the checks or do I have to push again?

de-nordic commented:
> @de-nordic the pipeline seems to be fixed could you maybe rerun the checks or do I have to push again?

I did re-run of failed jobs, but if think the PR needs to be rebased.

clamattia commented:
Is there anything left for this?
Is this PR just waiting on a second review?

de-nordic commented:
> Is there anything left for this? Is this PR just waiting on a second review?

Second reviewer. You can also review.

clamattia reviewed (approved):
Nice, thank you.

bjarki-andreasen reviewed (approved):
Nice work :)
